classes/robot/crawler.php

```diff
@@ -549,53 +551,58 @@ public function process_queue($verbose = false) {
+        while ($hastime) {
+            if (empty($nodes)) {
+                // Grab a list of items from the front of the queue
+                $nodes = $DB->get_records_sql('SELECT *
                                          FROM {tool_crawler_url}
                                         WHERE lastcrawled IS NULL
                                            OR lastcrawled < needscrawl
                                      ORDER BY priority DESC, needscrawl ASC, id ASC
-                                        LIMIT 1
-                                    ');
+                                        LIMIT :limitadhocworkers
+                                    ', ['limitadhocworkers' => 1000]);
+                if (empty($nodes)) {
+                    return $hastime; // The queue is empty
+                }
+            }
+            $node = array_shift($nodes);
+            $resource = (string)$node->id; // The node id is the unique resource that we want to lock on.
 
-            if ($config->uselogs == 1) {
+            // Get a new lock for the resource, wait for it if needed.
+            if (!$lock = $lockfactory->get_lock($resource, $timeout)) {
+                continue; // Try crawl the next node, this one is already being processed.
+            }
 
-                if (isset($node->courseid)) {
+            // If the course id is not in recent courses, remove it from the queue.
+            if ($config->uselogs == 1 && isset($node->courseid) && !in_array($node->courseid, $recentcourses)) {
+                // Will not show up in queue, but still keeps the data.
+                // in case the course becomes recently active in the future.
+                $node->needscrawl = $node->lastcrawled;
+                $DB->update_record('tool_crawler_url', $node);
+                $lock->release();
+                continue;
+            }
 
-                    // If the course id is not in recent courses, remove it from the queue.
-                    if (!in_array($node->courseid, $recentcourses)) {
+            $this->crawl($node, $verbose);
 
-                        // Will not show up in queue, but still keeps the data
-                        // in case the course becomes recently active in the future.
-                        $node->needscrawl = $node->lastcrawled;
-                        $DB->update_record('tool_crawler_url', $node);
-                    } else {
-                        break;
-                    }
-                } else {
-                    break;
-                }
-            } else {
-                break;
-            }
+            $lock->release();
 
             $hastime = time() < $cronstop;
             set_config('crawltick', time(), 'tool_crawler');
         }
-
-        if (isset($node) && $node !== false) {
-            $this->crawl($node, $verbose);
-            return true;
-        }
-
-        return false;
-
+        return $hastime;
```

Comment:
shouldnt this be hard coded to return true? ie there is more stuff to process on the next cron tick?

Reply:
we can hardcode it instead, will make it easier to read